### PR TITLE
Sequences

### PR DESCRIPTION
**Pull request by zianke**
Add support for nextval and currval functions.

### CODE REVIEW

File: src/catalog/database_catalog.cpp [hunk 1]
```diff
@@ -1327,6 +1353,240 @@ std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const Index
...
+  auto *const sequences_insert_pr = sequences_insert_redo->Delta();
+  // Write the sequence_oid into the PR
+  sequence_oid_offset = pg_sequence_all_cols_prm_[postgres::SEQRELID_COL_OID];
+  sequence_oid_ptr = sequences_insert_pr->AccessForceNotNull(sequence_oid_offset);
+  *(reinterpret_cast<sequence_oid_t *>(sequence_oid_ptr)) = sequence_oid;
+
+  // Write lastval into the PR. Default is 0
+  const auto sequence_lastval_offset = pg_sequence_all_cols_prm_[postgres::SEQLASTVAL_COL_OID];
+  const auto sequence_lastval_ptr = sequences_insert_pr->AccessForceNotNull(sequence_lastval_offset);
+  *(reinterpret_cast<uint32_t *>(sequence_lastval_ptr)) = 0;
```
Comment (YingjingLu) [inline src/catalog/database_catalog.cpp:1456]: Should we allow value to be passed in this entry for potential other start?

File: src/catalog/postgres/builder.cpp [hunk 2]
```diff
@@ -197,6 +203,33 @@ Schema Builder::GetClassTableSchema() {
...
+  columns.back().SetOid(SEQOID_COL_OID);
+
+  columns.emplace_back("seqrelid", type::TypeId::INTEGER, false, MakeNull(type::TypeId::INTEGER));
+  columns.back().SetOid(SEQRELID_COL_OID);
+
+  columns.emplace_back("seqlastval", type::TypeId::INTEGER, false, MakeNull(type::TypeId::INTEGER));
+  columns.back().SetOid(SEQLASTVAL_COL_OID);
+
+  return Schema(columns);
+}
```
Comment (YingjingLu) [inline src/catalog/postgres/builder.cpp:219]: Should we align the schema with the Postgres pg_sequence?

File: test/catalog/catalog_test.cpp [hunk 3]
```diff
@@ -254,6 +255,45 @@ TEST_F(CatalogTests, NamespaceTest) {
...
+  EXPECT_EQ(sequence_oid, catalog::INVALID_SEQUENCE_OID);
+  txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+
+  txn = txn_manager_->BeginTransaction();
+  accessor = catalog_->GetAccessor(common::ManagedPointer(txn), db_);
+  sequence_oid = accessor->GetSequenceOid("test_sequence");
+  EXPECT_EQ(sequence_oid, catalog::INVALID_SEQUENCE_OID);
+  txn_manager_->Abort(txn);
+}
+
```
Comment (YingjingLu) [inline test/catalog/catalog_test.cpp:296]: Would love to see a runtime increment test here. Especially when increment goes out of bound

File: script/testing/junit/src/SequenceTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        stmt = conn.createStatement();
+        ResultSet rs = stmt.executeQuery(select_SQL);
+        boolean exists = rs.next();
+        assert(exists);
+        checkIntRow(rs, new String [] {"nextval(VARCHAR)"}, new int [] {1});
+        assertNoMoreRows(rs);
+        String drop_SQL = "DROP SEQUENCE seq;";
+        stmt = conn.createStatement();
+        stmt.execute(drop_SQL);
+    }
```
Comment (wuwenw) [inline script/testing/junit/src/SequenceTest.java:83]: more query for junit test would be nice

File: src/catalog/database_catalog.cpp [hunk 5]
```diff
@@ -1327,6 +1353,240 @@ std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const Index
...
+      table_pr->AccessForceNotNull(pg_class_all_cols_prm_[postgres::RELNAMESPACE_COL_OID])));
+  const storage::VarlenEntry name_varlen = *(reinterpret_cast<const storage::VarlenEntry *const>(
+      table_pr->AccessForceNotNull(pg_class_all_cols_prm_[postgres::RELNAME_COL_OID])));
+
+  // TODO(zianke): schema_ptr and sequence_ptr currently empty
+  // auto *const schema_ptr = *(reinterpret_cast<const IndexSchema *const *const>(
+  //      table_pr->AccessForceNotNull(pg_class_all_cols_prm_[postgres::REL_SCHEMA_COL_OID])));
+  // auto *const sequence_ptr = *(reinterpret_cast<storage::index::Index *const *const>(
+  //      table_pr->AccessForceNotNull(pg_class_all_cols_prm_[postgres::REL_PTR_COL_OID])));
+
```
Comment (wuwenw) [inline src/catalog/database_catalog.cpp:1526]: Is it going to be implemented in the future?

File: src/execution/sql/functions/string_functions.cpp [hunk 6]
```diff
@@ -213,56 +213,90 @@ void StringFunctions::Nextval(exec::ExecutionContext *ctx, Integer *result, cons
...
+    auto temp_table_oid = ctx->GetTempTable();
+    auto temp_table = accessor->GetTable(temp_table_oid).Get();
+    auto temp_colums = accessor->GetSchema(temp_table_oid).GetColumns();
+
+
+    // Sequence table only have two colums right now
+    const std::vector<catalog::col_oid_t> temp_colums_oids{temp_colums[0].Oid(), temp_colums[1].Oid()};
+    auto const table_projection_map =  temp_table->ProjectionMapForOids(temp_colums_oids);
+
+    const auto pci = temp_table->InitializerForProjectedColumns(temp_colums_oids, 100);
```
Comment (parapluirevel) [inline src/execution/sql/functions/string_functions.cpp:313]: Optional: define this value as a constant.

File: src/execution/sql/functions/string_functions.cpp [hunk 7]
```diff
@@ -213,56 +213,90 @@ void StringFunctions::Nextval(exec::ExecutionContext *ctx, Integer *result, cons
...
+
+    auto sequence_oid = accessor->GetSequenceOid(s);
+    //common::ManagedPointer<SequenceMetadata> seq = accessor->GetSequence(sequence_oid);
+
+    auto temp_table_oid = ctx->GetTempTable();
+    auto temp_table = accessor->GetTable(temp_table_oid).Get();
+    auto temp_colums = accessor->GetSchema(temp_table_oid).GetColumns();
+
+
+    // Sequence table only have two colums right now
```
Comment (parapluirevel) [inline src/execution/sql/functions/string_functions.cpp:308]: columns

File: src/execution/sql/functions/string_functions.cpp [hunk 8]
```diff
@@ -213,56 +213,90 @@ void StringFunctions::Nextval(exec::ExecutionContext *ctx, Integer *result, cons
...
+    const std::vector<catalog::col_oid_t> temp_colums_oids{temp_colums[0].Oid(), temp_colums[1].Oid()};
+    auto const table_projection_map =  temp_table->ProjectionMapForOids(temp_colums_oids);
```
Comment (parapluirevel) [inline src/execution/sql/functions/string_functions.cpp:310]: extra space after "="

File: src/network/postgres/postgres_protocol_interpreter.cpp [hunk 9]
```diff
@@ -128,11 +128,12 @@ Transition PostgresProtocolInterpreter::ProcessStartup(const common::ManagedPoin
     std::this_thread::sleep_for(std::chrono::milliseconds{sleep_time});
     sleep_time *= BACKOFF_FACTOR;
   } while (sleep_time <= MAX_BACKOFF_TIME);
+
   if (temp_oid == catalog::INVALID_TABLE_OID) {
-    STORAGE_LOG_ERROR("INVALID !!!");
-    writer.WriteErrorResponse(
-        "ERROR:  Failed to create a temporary table for this connection.");
-    return Transition::TERMINATE;
+      STORAGE_LOG_ERROR("INVALID !!!");
```
Comment (parapluirevel) [inline src/network/postgres/postgres_protocol_interpreter.cpp:133]: Make sure the indentation is consistent with the codebase (here and elsewhere)

File: src/network/postgres/postgres_protocol_interpreter.cpp [hunk 10]
```diff
@@ -128,11 +128,12 @@ Transition PostgresProtocolInterpreter::ProcessStartup(const common::ManagedPoin
...
+      writer.WriteErrorResponse(
+              "ERROR:  Failed to create a temporary table for this connection.");
+      return Transition::TERMINATE;
```
Comment (parapluirevel) [inline src/network/postgres/postgres_protocol_interpreter.cpp:136]: Make sure the indentation is consistent with the codebase's (here and elsewhere)

File: src/catalog/database_catalog.cpp [hunk 11]
```diff
@@ -1327,6 +1353,240 @@ std::vector<std::pair<common::ManagedPointer<storage::index::Index>, const Index
...
+  }
+
+  // Insert into namespace_index
+  index_pr = class_ns_index_init.InitializeRow(index_buffer);
+  *(reinterpret_cast<namespace_oid_t *>(index_pr->AccessForceNotNull(0))) = ns_oid;
+  const auto result UNUSED_ATTRIBUTE = classes_namespace_index_->Insert(txn, *index_pr, class_tuple_slot);
+  TERRIER_ASSERT(result, "Insertion into non-unique namespace index failed.");
+
+  // TODO(zianke): Next, insert sequence metadata into pg_sequence
+
```
Comment (wuwenw) [inline src/catalog/database_catalog.cpp:1445]: Why does pg_sequence only have two attributes?

File: src/catalog/catalog_accessor.cpp [hunk 12]
```diff
@@ -133,6 +134,27 @@ common::ManagedPointer<storage::index::Index> CatalogAccessor::GetIndex(index_oi
...
+    if (search_result != INVALID_SEQUENCE_OID) return search_result;
+  }
+  return INVALID_SEQUENCE_OID;
+}
+
+sequence_oid_t CatalogAccessor::GetSequenceOid(namespace_oid_t ns, std::string name) const {
+  NormalizeObjectName(&name);
+  return dbc_->GetSequenceOid(txn_, ns, name);
+}
+
```
Comment (wuwenw) [inline src/catalog/catalog_accessor.cpp:155]: What's the difference between these two GetSequenceOid()?

File: src/traffic_cop/traffic_cop.cpp [hunk 13]
```diff
@@ -223,6 +233,13 @@ TrafficCopResult TrafficCop::ExecuteDropStatement(
       }
       break;
     }
+    case network::QueryType::QUERY_DROP_SEQUENCE: {
+      if (execution::sql::DDLExecutors::DropSequenceExecutor(
```
Comment (wuwenw) [inline src/traffic_cop/traffic_cop.cpp:237]: also need a check in drop table for cascade

**Comment by wuwenw** [comment]
Overall greate work with comprehensive test coverage, though with minor format issue. Pg_sequence can be aligned with postgres for more functionalities in the future, and it would be nice if checks are added for DROP TABLE ... CASCADE.

**Comment by zianke** [comment]
Please refer to the [Final Code Drop](https://github.com/cmu-db/terrier/pull/897).

**Comment by zianke** [comment]
Please refer to the [Final Code Drop](https://github.com/cmu-db/terrier/pull/897).

**Pull request closed without merging**